cmd/copyEnumeratorInit.go

```diff
@@ -0,0 +1,185 @@
+package cmd
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+
+	"github.com/Azure/azure-storage-azcopy/common"
+)
+
+func (cca *cookedCopyCmdArgs) initEnumerator(jobPartOrder common.CopyJobPartOrderRequest, ctx context.Context) (*copyEnumerator, error) {
+	var traverser resourceTraverser
+
+	dst := cca.destination
+
+	if cca.destinationSAS != "" {
+		destURL, err := url.Parse(dst)
+
+		if err != nil {
+			return nil, err
+		}
+
+		destURL = copyHandlerUtil{}.appendQueryParamToUrl(destURL, cca.destinationSAS)
+		dst = destURL.String()
+	}
+
+	// TODO: in download refactor, handle trailing wildcard properly here.
+	// As is, we don't cut it off at the moment,
+	// and we also don't properly handle the "source points to contents" scenario aside from on local, which waives it through wildcard support.
+	traverser, err := initResourceTraverser(cca.source, cca.fromTo.From(), nil, nil, &cca.followSymlinks, cca.listOfFilesChannel, cca.recursive, true, func() {})
+	if err != nil {
+		return nil, err
+	}
+
+	// source directory check occurs within the traversers. This allows us to error out properly when pointing at a folder.
```

Comment:
The check on the source was moved to the traverser because I'd catch a lot of exceptions with when a transfer would actually succeed on the download refactor. This actually ensured the failure would occur on a directory.

Realistically, I feel as though it might be a little unsafe to implement it any other way. resourceTraversers still have .isDirectory, and we could pretty safely only use that on source with all checks needed, as we currently have the isDestDirectory function.

I did rethink it though, and I'm going to retry the isDirectory approach compared to this setup.